frontend/src/components/editor/app-container.tsx

```diff
@@ -20,6 +27,52 @@ export const AppContainer: React.FC<PropsWithChildren<Props>> = ({
   isRunning,
   children,
 }) => {
+  const errors = useCellErrors();
+
+  // Dynamically update favicon for run feedback
+  let favicon: HTMLLinkElement | null =
+    document.querySelector("link[rel~='icon']");
+
+  if (!favicon) {
+    favicon = document.createElement("link");
+    favicon.rel = "icon";
+    document.getElementsByTagName("head")[0].append(favicon);
+  }
+
+  const favicons: Record<string, string> = useMemo(
+    () => ({
+      idle: "./favicon.ico",
+      success: "./circle-check.ico",
+      running: "./circle-play.ico",
+      error: "./circle-x.ico",
+    }),
+    [],
+  );
+
+  const resetFaviconIfComplete = useCallback(() => {
+    if (!isRunning && document.visibilityState == "visible") {
+      setTimeout(() => {
+        favicon.href = favicons.idle;
+      }, 3000);
+    }
+  }, [favicon, favicons.idle, isRunning]);
+
+  useEffect(() => {
+    if (isRunning) {
+      favicon.href = favicons.running;
+      return;
+    }
+    favicon.href = errors.length > 0 ? favicons.error : favicons.success;
+    resetFaviconIfComplete();
+    return () => {
+      favicon.href = favicons.idle;
+    };
+  }, [isRunning, errors, favicon, favicons, resetFaviconIfComplete]);
```

Comment:
we can probably inline `resetFaviconIfComplete`, because we will want to clear the timeout too on the returned `unsubscribe`

```ts
  useEffect(() => {
    if (isRunning) {
      FAVICONS.href = favicons.running;
      return;
    }
    favicon.href = errors.length > 0 ? FAVICONS.error : FAVICONS.success;
    // Reset back to idle after 3 seconds
    const timeout = setTimeout(() => {
        favicon.href = FAVICONS.idle;
      }, 3000);
    return () => {
      favicon.href = FAVICONS.idle;
      clearTimeout();
    };
  }, [isRunning, errors.length]);
```

Reply:
Good catch thanks!